kbs/src/resource/plugin/nebula.rs

```diff
@@ -0,0 +1,462 @@
+        let key: PathBuf = temp_dir.path().join(KEY_FILENAME);
+
+        Ok(Self {
+            _temp_dir: temp_dir,
+            crt,
+            key,
+        })
+    }
+
+    /// Run "nebula-cert sign" to generate a credential
+    pub fn generate(
+        &self,
+        ca_key: &Path,
+        ca_crt: &Path,
+        params: &NebulaCredentialParams,
+    ) -> Result<&Self> {
+        let mut args: Vec<OsString> = Vec::from(params);
+
+        args.push("-ca-key".into());
+        args.push(ca_key.into());
+        args.push("-ca-crt".into());
+        args.push(ca_crt.into());
+        args.push("-out-key".into());
+        args.push(self.key.as_path().into());
+        args.push("-out-crt".into());
+        args.push(self.crt.as_path().into());
+
+        let status = Command::new(NEBULA_CERT_BIN)
+            .args(args)
+            .status()
+            .context("nebula-cert sign run")?;
+
+        if !status.success() {
+            bail!("nebula-cert sign status");
+        }
+
+        Ok(self)
+    }
+}
+
+/// The temp_dir is auto-deleted when it goes out-of-scope, but before that
+/// we need to delete the generated credential
+impl Drop for Credential {
+    fn drop(&mut self) {
+        if self.crt.exists() {
+            if let Err(e) = fs::remove_file(self.crt.as_path())
+                .with_context(|| format!("Remove {} file", self.crt.display()))
+            {
+                log::warn!("{}", e.to_string());
+            }
+        }
+        if self.key.exists() {
+            if let Err(e) = fs::remove_file(self.key.as_path())
+                .with_context(|| format!("Remove {} file", self.key.display()))
+            {
+                log::warn!("{}", e.to_string());
+            }
+        }
+    }
+}
+
+/// Nebula Certificate Authority
+#[derive(Debug, Default)]
+struct NebulaCa {
+    key: PathBuf,
+    crt: PathBuf,
+    work_dir: PathBuf,
+}
+
+impl NebulaCa {
+    pub fn get_credential_resource(&self, params: &NebulaCredentialParams) -> Result<Vec<u8>> {
+        let cred = Credential::new(self.work_dir.as_path())?;
+
+        cred.generate(self.key.as_path(), self.crt.as_path(), params)?;
+
+        let resource = CredentialResource {
+            node_crt: fs::read(cred.crt.as_path())
+                .with_context(|| format!("read {}", cred.crt.display()))?,
+            node_key: fs::read(cred.key.as_path())
+                .with_context(|| format!("read {}", cred.key.display()))?,
+            ca_crt: fs::read(self.crt.as_path())
+                .with_context(|| format!("read {}", self.crt.display()))?,
+        };
+
+        Ok(serde_json::to_vec(&resource)?)
+    }
+
+    pub fn get_version(&self) -> Result<String> {
+        let output = Command::new(NEBULA_CERT_BIN).arg("--version").output()?;
+        Ok(String::from_utf8(output.stdout)?)
+    }
+
+    pub fn test_all(&self) -> Result<()> {
+        self.test_nebula_cert_sign()
+    }
```

Comment:
Do you call this test anywhere?

It's pretty common to put test code in a separate module rather than implementing it alongside the real methods. Not sure it's a big deal.

Reply:
It is called when the plugin is initialized. It runs simple tests to check if the provided `nebula-cert` is working.